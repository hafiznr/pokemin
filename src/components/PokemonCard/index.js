/** @jsxRuntime classic */
/** @jsx jsx */
import { css, jsx } from '@emotion/react';
import { useHistory } from "react-router-dom";
import { Badge, Box, Button } from "@chakra-ui/react";
import { getPokemonNumber } from '../../utils';

const PokemonCard = ({
  id, 
  name, 
  nickname, 
  image, 
  isMyPokemon,
  owned,
  onRelease
}) => {
  const history = useHistory();

  const imgStyles = css({
    backgroundColor: '#fff',
    borderRadius: '8px'
  })

  const parsedId = getPokemonNumber(id);

  const handleClickRelease = (e) => {
    e.stopPropagation();
    
    onRelease({
      id, 
      name,
      nickname, 
      image
    });
  }

  const handleClickCard = () => {
    history.push(`/pokemon/${name}`)
  }

  return (
      <Box 
        d="grid" 
        bg="gray.100" 
        borderRadius="lg" 
        alignItems="center" 
        mb="2" 
        gridGap={4}
        gridTemplateColumns="96px 1fr"
        padding="8px"
        position="relative"
        onClick={handleClickCard}
      >
        <img 
          src={image} 
          alt={nickname || name} 
          loading="lazy" 
          width="96px" 
          height="96px"
          css={imgStyles}
        />
        <Box>
          <Badge 
            fontSize=""
            borderRadius="full" 
            colorScheme="blackAlpha"
          >
            #{parsedId}
          </Badge>
          <Box
            margin="2px 0"
            fontWeight="semibold" 
            fontSize="2xl"
            textTransform="capitalize"
          >
            {nickname || name}
          </Box>
          {isMyPokemon ? (
            <Button 
              colorScheme="red" 
              mt="4px" 
              size="md"
              position="absolute"
              right="8px"
              bottom="8px"
              onClick={handleClickRelease}
            >
              Release
            </Button>
          ) : (
            <Badge 
              fontSize="md"
              borderRadius="full" 
              colorScheme={owned ? "telegram" : "yellow"}
            >
              {owned ? `Owned: ${owned}` : 'Not owned'}
            </Badge>
          )}
        </Box>
      </Box>
  )
}

export default PokemonCard;