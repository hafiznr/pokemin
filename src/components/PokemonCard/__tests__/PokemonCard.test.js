import { render, screen, fireEvent } from '@testing-library/react';
import PokemonCard from '../';

const onRelease = jest.fn();

jest.mock("react-router-dom", () => ({
  ...jest.requireActual("react-router-dom"),
  useHistory: () => ({
    push: () => {}
  })
}));

test('renders PokemonCard', () => {
  const { container, rerender } = render(<PokemonCard/>);

  expect(screen.getByTestId('owned-badge')).toHaveTextContent('Not owned');
  expect(container).toMatchSnapshot();

  rerender(<PokemonCard owned={3}/>)
  expect(screen.getByTestId('owned-badge')).toHaveTextContent('Owned: 3');

  rerender(<PokemonCard isMyPokemon nickname="hihi" onRelease={onRelease}/>)
  expect(screen.getByTestId('release-button-hihi')).toBeInTheDocument();
  
  fireEvent.click(screen.getByText(/Release/i));
  expect(onRelease).toHaveBeenCalledTimes(1);

  fireEvent.click(screen.getByTestId('pokemon-card'));
});