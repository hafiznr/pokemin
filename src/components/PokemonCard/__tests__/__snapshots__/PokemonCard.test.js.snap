// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders PokemonCard 1`] = `
<div>
  <div
    class="css-gh77c9"
    data-testid="pokemon-card"
  >
    <img
      class="css-1yjt0aa-PokemonCard"
      height="96px"
      loading="lazy"
      width="96px"
    />
    <div
      class="css-0"
    >
      <span
        class="chakra-badge css-uwa4md"
      >
        #
      </span>
      <div
        class="css-1wjjfwl"
      />
      <span
        class="chakra-badge css-1lyftqe"
        data-testid="owned-badge"
      >
        Not owned
      </span>
    </div>
  </div>
</div>
`;
